File: epi_judge_cpp_solutions/two_thread_increment.cc

#include <iostream>
#include <string>
#include <thread>

using std::cout;
using std::endl;
using std::stoi;
using std::thread;

static int counter = 0;

void IncrementThread(int n) {
  for (int i = 0; i < n; ++i) {
    ++counter;
  }
}

void TwoThreadIncrementDriver(int n) {
  thread T1(IncrementThread, n);
  thread T2(IncrementThread, n);
  T1.join();
  T2.join();

  cout << counter << endl;
}

int main(int argc, char* argv[]) {
  int n = argc == 2 ? stoi(argv[1]) : 1000000000;
  TwoThreadIncrementDriver(n);
  return 0;
}

File: epi_judge_python_solutions/two_thread_increment.py
import sys
import threading

counter = 0


def increment_thread(n):
    global counter
    for _ in range(n):
        counter = counter + 1


def two_thread_increment_driver(n):
    t1 = threading.Thread(target=increment_thread, args=(n, ))
    t2 = threading.Thread(target=increment_thread, args=(n, ))

    t1.start()
    t2.start()
    t1.join()
    t2.join()

    print(counter)


if __name__ == '__main__':
    n = int(sys.argv[1]) if len(sys.argv) == 2 else 1000000000
    two_thread_increment_driver(n)
